# Distances Between Observations

Read this notebook from top to bottom and fill in the code as you go. Work together and discuss with other students in the class. Try to resolve any errors on your own first, but don't get stuck; ask for help!

In addition to writing and running code, be sure to examine any output and interpret the results before moving on.

For many of these questions, there are several approaches, and there is no single right answer. You should try a few different things and compare with your classmates.

We will use `scikit-learn` extensively later, but for this activity you might want to stick with `pandas`.


In [61]:
import pandas as pd
import numpy as np

## Ames - Recommending Similar Homes

1\. Suppose that you really like house 0 in the Ames housing data set, but it is too expensive. Find cheaper homes that are similar to it --- in terms of living area, number of bedrooms, number of bathrooms --- by calculating distances from house 0. You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar (based on these 3 variables). Do they make sense?

_Think:_ If the goal is to find a "good deal" on a similar house, should sale price be included as a variable in your distance metric?

In [62]:
df_ames = pd.read_csv("https://raw.githubusercontent.com/kevindavisross/data301/main/data/AmesHousing.txt", sep="\t")

In [63]:
# YOUR CODE HERE. ADD CELLS AS NEEDED
df_ames["Bathrooms"] = df_ames["Full Bath"] + df_ames["Bsmt Full Bath"]

house0 = df_ames.loc[0]

show_vars = ["Gr Liv Area", "Bedroom AbvGr", "Bathrooms", "House Style", "Year Built", "SalePrice"]

house0[show_vars]

Gr Liv Area        1656
Bedroom AbvGr         3
Bathrooms           2.0
House Style      1Story
Year Built         1960
SalePrice        215000
Name: 0, dtype: object

In [64]:
features = ["Gr Liv Area", "Bedroom AbvGr", "Bathrooms"]
X = df_ames[features]. astype(float)

X_z = (X - X.mean()) / X.std()

diff = X_z -X_z.loc[0]

df_ames["dist_euclid"] = np.sqrt((diff) **2).sum(1)
df_ames["dist_manhattan"] = diff.abs().sum(axis =1)

In [65]:
df_cheaper_sales_price = df_ames[df_ames["SalePrice"]<house0["SalePrice"]]

df_cheaper_sales_price.sort_values("dist_euclid")[show_vars]

,Gr Liv Area,Bedroom AbvGr,Bathrooms,House Style,Year Built,SalePrice
655,1656,3,2.0,1.5Fin,1916,135000
1493,1656,3,2.0,2Story,2003,165400
1927,1657,3,2.0,1Story,1970,163500
1142,1657,3,2.0,2Story,1994,189000
10,1655,3,2.0,2Story,1993,175900
...,...,...,...,...,...,...
2195,2592,6,4.0,2Story,1923,150000
2181,4676,3,4.0,2Story,2007,184750
2180,5095,2,3.0,1Story,2008,183850
2194,3395,8,2.0,2.5Fin,1914,200000


In [66]:
df_cheaper_sales_price.sort_values("dist_euclid")[show_vars].head(10)

,Gr Liv Area,Bedroom AbvGr,Bathrooms,House Style,Year Built,SalePrice
655,1656,3,2.0,1.5Fin,1916,135000
1493,1656,3,2.0,2Story,2003,165400
1927,1657,3,2.0,1Story,1970,163500
1142,1657,3,2.0,2Story,1994,189000
10,1655,3,2.0,2Story,1993,175900
731,1654,3,2.0,1.5Fin,1930,137000
73,1659,3,2.0,2Story,2004,181000
1895,1652,3,2.0,1Story,1959,200000
1436,1652,3,2.0,2Story,2000,207500
846,1660,3,2.0,2Story,2002,195000


In [67]:
compare_vars = ["Gr Liv Area", "Bedroom AbvGr", "Bathrooms"]

df_scaled = (
    df_ames[compare_vars] - df_ames[compare_vars].mean()
) / df_ames[compare_vars].std()

In [68]:
df_ames["dist_scaled"] = np.sqrt(
    ((df_scaled - df_scaled.loc[0]) ** 2).sum(axis=1)
)

In [69]:
df_cheaper_sales_price = df_ames[
    df_ames["SalePrice"] < house0["SalePrice"]
]

In [70]:
show_vars_scaled = [
    "SalePrice",
    "Gr Liv Area",
    "Bedroom AbvGr",
    "Bathrooms",
    "dist_scaled"
]

df_cheaper_sales_price.sort_values("dist_scaled")[show_vars_scaled].head(10)

,SalePrice,Gr Liv Area,Bedroom AbvGr,Bathrooms,dist_scaled
655,135000,1656,3,2.0,0.000000
1493,165400,1656,3,2.0,0.000000
1927,163500,1657,3,2.0,0.001978
1142,189000,1657,3,2.0,0.001978
10,175900,1655,3,2.0,0.001978
731,137000,1654,3,2.0,0.003956
73,181000,1659,3,2.0,0.005935
1895,200000,1652,3,2.0,0.007913
1436,207500,1652,3,2.0,0.007913
846,195000,1660,3,2.0,0.007913


In [71]:
compare_vars = ["Gr Liv Area", "Bedroom AbvGr", "Bathrooms"]

df_scaled = (
    df_ames[compare_vars] - df_ames[compare_vars].mean()
) / df_ames[compare_vars].std()

In [72]:
df_ames["dist_scaled"] = np.sqrt(
    ((df_scaled - df_scaled.loc[0]) ** 2).sum(axis=1)
)

In [73]:
df_cheaper_sales_price = df_ames[
    df_ames["SalePrice"] < house0["SalePrice"]
]

show_vars_scaled = [
    "SalePrice",
    "Gr Liv Area",
    "Bedroom AbvGr",
    "Bathrooms",
    "dist_scaled"
]

df_cheaper_sales_price.sort_values("dist_scaled")[show_vars_scaled].head(10)

,SalePrice,Gr Liv Area,Bedroom AbvGr,Bathrooms,dist_scaled
655,135000,1656,3,2.0,0.000000
1493,165400,1656,3,2.0,0.000000
1927,163500,1657,3,2.0,0.001978
1142,189000,1657,3,2.0,0.001978
10,175900,1655,3,2.0,0.001978
731,137000,1654,3,2.0,0.003956
73,181000,1659,3,2.0,0.005935
1895,200000,1652,3,2.0,0.007913
1436,207500,1652,3,2.0,0.007913
846,195000,1660,3,2.0,0.007913


**YOUR RESPONSE HERE**

The similar homes make sense because they have smiliar living around 1,652-1660sq.ft, 3 bedrooms, and similiar bathrooms. Scaling did not change the results much. Sale price should not be included because we want to find similar homes that are cheaper 

2\. Continuing part 1. Suppose that you really like house 0 in the data set, but it is too expensive. Find cheaper homes that are similar to it --- in terms of living area, number of bedrooms, number of bathrooms, **and House Style** --- by calculating distances from house 0. You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar. Do they make sense?

In [74]:
# YOUR CODE HERE. ADD CELLS AS NEEDED
house_style_dummies = pd.get_dummies(df_ames["House Style"], prefix="Style", dtype=int)

df_ames2 = pd.concat([df_ames, house_style_dummies], axis=1)

In [75]:
compare_vars = ["Gr Liv Area", "Bedroom AbvGr", "Bathrooms"] + list(house_style_dummies.columns)

In [76]:
df_scaled2 = (
    df_ames2[compare_vars] - df_ames2[compare_vars].mean()
) / df_ames2[compare_vars].std()

In [77]:
df_ames2["dist_style"] = np.sqrt(
    ((df_scaled2 - df_scaled2.loc[0]) ** 2).sum(axis=1)
)

In [78]:
cheaper2 = df_ames2[
    df_ames2["SalePrice"] < house0["SalePrice"]
]

In [79]:
show_vars2 = [
    "SalePrice",
    "Gr Liv Area",
    "Bedroom AbvGr",
    "Bathrooms",
    "House Style",
    "dist_style"
]

cheaper2.sort_values("dist_style")[show_vars2].head(10)

,SalePrice,Gr Liv Area,Bedroom AbvGr,Bathrooms,House Style,dist_style
1927,163500,1657,3,2.0,1Story,0.001978
1895,200000,1652,3,2.0,1Story,0.007913
1940,153000,1647,3,2.0,1Story,0.017804
618,167000,1644,3,2.0,1Story,0.023738
375,200000,1671,3,2.0,1Story,0.029673
209,173000,1675,3,2.0,1Story,0.037586
2294,196000,1676,3,2.0,1Story,0.039564
876,213000,1630,3,2.0,1Story,0.051433
1185,174000,1682,3,2.0,1Story,0.051433
2075,130000,1630,3,2.0,1Story,0.051433


**YOUR RESPONSE HERE**
Adding House Style changed the closest matches, showing that the results are sensitive to the variables included. The selected homes still have similar living areas and 3 bedrooms, but now house style is also considered, making the recommendations more specific.


3\. Continuing parts 1 and 2. Suppose that you really like house 0 in the data set, but it is too expensive. Find cheaper homes that are similar to it, by calculating distances. You can **choose the variables to include, but include both quantitative and categorical variables**. Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar. Do they make sense?

You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

_Hint:_ There are many variables in the data set. Do not attempt to compute distance based on all the variables! You will want to pare down the number of variables, but be sure to include a mixture of categorical and quantitative variables. Refer to the [data documentation](https://ww2.amstat.org/publications/jse/v19n3/decock/DataDocumentation.txt) for information about the variables.


In [80]:
# YOUR CODE HERE. ADD CELLS AS NEEDED
cat_dummies = pd.get_dummies(
    df_ames[["House Style", "Neighborhood"]],
    dtype=int
)

df_ames3 = pd.concat([df_ames, cat_dummies], axis=1)

In [81]:
quant_vars = [
    "Gr Liv Area",
    "Bedroom AbvGr",
    "Bathrooms",
    "Overall Qual"
]

In [82]:
compare_vars3 = quant_vars + list(cat_dummies.columns)

In [83]:
df_scaled3 = (
    df_ames3[compare_vars3] - df_ames3[compare_vars3].mean()
) / df_ames3[compare_vars3].std()

In [84]:
df_ames3["dist_final"] = np.sqrt(
    ((df_scaled3 - df_scaled3.loc[0]) ** 2).sum(axis=1)
)

In [85]:
cheaper3 = df_ames3[
    df_ames3["SalePrice"] < df_ames3.loc[0, "SalePrice"]
]

In [86]:
show_vars3 = [
    "SalePrice",
    "Gr Liv Area",
    "Bedroom AbvGr",
    "Bathrooms",
    "Overall Qual",
    "House Style",
    "Neighborhood",
    "dist_final"
]

cheaper3.sort_values("dist_final")[show_vars3].head(10)

,SalePrice,Gr Liv Area,Bedroom AbvGr,Bathrooms,Overall Qual,House Style,Neighborhood,dist_final
1927,163500,1657,3,2.0,6,1Story,NAmes,0.001978
1895,200000,1652,3,2.0,6,1Story,NAmes,0.007913
618,167000,1644,3,2.0,6,1Story,NAmes,0.023738
1611,167300,1587,3,2.0,6,1Story,NAmes,0.136496
147,159500,1580,3,2.0,6,1Story,NAmes,0.150344
1240,166800,1570,3,2.0,6,1Story,NAmes,0.170126
1231,174000,1537,3,2.0,6,1Story,NAmes,0.235406
1248,180000,1790,3,2.0,6,1Story,NAmes,0.265079
2546,167000,1488,3,2.0,6,1Story,NAmes,0.332338
631,156500,1478,3,2.0,6,1Story,NAmes,0.352120


**YOUR RESPONSE HERE**

The closest homes make sense because they have similar living area, bedrooms, bathrooms, quality, house style, and neighborhood. Adding more quantitative and categorical variables changed some of the closest matches, showing that the results are sensitive to the variables and scaling method chosen.

## Colleges similar to Cal Poly

We'll use data from the [College Scorecard data](https://collegescorecard.ed.gov/) to find colleges and universities that are similar to Cal Poly.

In [87]:
df_college = pd.read_csv("https://datasci112.stanford.edu/data/college_attributes.csv")

df_college.set_index("Institution", inplace = True)

df_college

,City,State,AdmissionRate,Undergraduates,CarnegieClassification,Ownership,PCIP01,PCIP03,PCIP04,PCIP05,...,PCIP44,PCIP45,PCIP46,PCIP47,PCIP48,PCIP49,PCIP50,PCIP51,PCIP52,PCIP54
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,Normal,AL,0.7160,5098.0,Master's Colleges & Universities: Larger Programs,Public,0.0445,0.0071,0.0053,0.0000,...,0.0409,0.0249,0.0,0.0,0.0,0.0,0.0231,0.0000,0.1637,0.0000
University of Alabama at Birmingham,Birmingham,AL,0.8854,13284.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0000,0.0000,0.0020,...,0.0195,0.0239,0.0,0.0,0.0,0.0,0.0249,0.2088,0.2159,0.0141
University of Alabama in Huntsville,Huntsville,AL,0.7367,7358.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0127,0.0,0.0,0.0,0.0,0.0407,0.1341,0.1930,0.0073
Alabama State University,Montgomery,AL,0.9799,3495.0,Doctoral/Professional Universities,Public,0.0000,0.0000,0.0000,0.0000,...,0.0648,0.0196,0.0,0.0,0.0,0.0,0.0511,0.0904,0.1513,0.0059
The University of Alabama,Tuscaloosa,AL,0.7890,30725.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0061,0.0000,0.0019,...,0.0072,0.0661,0.0,0.0,0.0,0.0,0.0234,0.1077,0.2916,0.0096
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,Miramar,FL,0.3088,239.0,Not applicable,Private for-profit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,1.0000,0.0000,0.0000
Herzing University-Tampa,Tampa,FL,0.9630,68.0,Not applicable,Private nonprofit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000
Abilene Christian University-Undergraduate Online,Addison,TX,1.0000,415.0,Not applicable,Private nonprofit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000


We'll want to single out Cal Poly, which we can do like this.

In [88]:
school_name = "California Polytechnic State University-San Luis Obispo"

cp = df_college.loc[school_name]

cp

City                                                        San Luis Obispo
State                                                                    CA
AdmissionRate                                                          0.33
Undergraduates                                                      21090.0
CarnegieClassification    Master's Colleges & Universities: Larger Programs
Ownership                                                            Public
PCIP01                                                               0.1084
PCIP03                                                               0.0255
PCIP04                                                               0.0441
PCIP05                                                               0.0019
PCIP09                                                               0.0353
PCIP10                                                               0.0175
PCIP11                                                               0.0326
PCIP12      

1\. Based on only the admission rate and the number of undergraduates, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [89]:
# YOUR CODE HERE. ADD CELLS AS NEEDED
df_college.columns



Index(['City', 'State', 'AdmissionRate', 'Undergraduates',
       'CarnegieClassification', 'Ownership', 'PCIP01', 'PCIP03', 'PCIP04',
       'PCIP05', 'PCIP09', 'PCIP10', 'PCIP11', 'PCIP12', 'PCIP13', 'PCIP14',
       'PCIP15', 'PCIP16', 'PCIP19', 'PCIP22', 'PCIP23', 'PCIP24', 'PCIP25',
       'PCIP26', 'PCIP27', 'PCIP29', 'PCIP30', 'PCIP31', 'PCIP38', 'PCIP39',
       'PCIP40', 'PCIP41', 'PCIP42', 'PCIP43', 'PCIP44', 'PCIP45', 'PCIP46',
       'PCIP47', 'PCIP48', 'PCIP49', 'PCIP50', 'PCIP51', 'PCIP52', 'PCIP54'],
      dtype='object')

In [90]:
compare_vars = ["AdmissionRate", "Undergraduates"]

df_college[compare_vars] = df_college[compare_vars].apply(
    pd.to_numeric, errors="coerce"
)

In [91]:
college_scaled = (
    df_college[compare_vars] - df_college[compare_vars].mean()
) / df_college[compare_vars].std()

cp_scaled = college_scaled.loc[school_name]

In [92]:
df_college["distance"] = np.sqrt(
    ((college_scaled - cp_scaled) ** 2).sum(axis=1)
)

In [93]:
df_college.sort_values("distance")[
    ["AdmissionRate", "Undergraduates", "distance"]
].head(11)

,AdmissionRate,Undergraduates,distance
Institution,,,
California Polytechnic State University-San Luis Obispo,0.3300,21090.0,0.000000
University of California-Santa Barbara,0.2918,23081.0,0.309162
DeVry University-Illinois,0.4552,19729.0,0.593121
University of North Carolina at Chapel Hill,0.2040,19722.0,0.596846
Clemson University,0.4922,21577.0,0.736788
University of Virginia-Main Campus,0.2074,17041.0,0.761296
CUNY Hunter College,0.4590,17293.0,0.761441
Stony Brook University,0.4806,17900.0,0.795756
Boston University,0.1865,17501.0,0.797041


**YOUR RESPONSE HERE**

I standardized admission rate and number of undergraduates, then used Euclidean distance to measure similarity to Cal Poly. The schools with the smallest distances were considered most similar. The closest match had a distance of 0.3092, meaning its admission rate and undergraduate enrollment were most similar to Cal Poly among the other schools.

2\. Now consider the admission rate, the number of undergraduates, and also the [Carnegie classification](https://en.wikipedia.org/wiki/Carnegie_Classification_of_Institutions_of_Higher_Education) of the type of school, and the ownership (public, private, etc.) Based on these variables, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [94]:
# YOUR CODE HERE. ADD CELLS AS NEEDED
cat_dummies = pd.get_dummies(
    df_college[["CarnegieClassification", "Ownership"]],
    dtype=int
)

df_college2 = pd.concat([df_college, cat_dummies], axis=1)

In [95]:
quant_vars = ["AdmissionRate", "Undergraduates"]

compare_vars2 = quant_vars + list(cat_dummies.columns)

In [96]:
college_scaled2 = (
    df_college2[compare_vars2] - df_college2[compare_vars2].mean()
) / df_college2[compare_vars2].std()

In [97]:
cp_scaled2 = college_scaled2.loc[school_name]

In [98]:
df_college2["distance2"] = np.sqrt(
    ((college_scaled2 - cp_scaled2) ** 2).sum(axis=1)
)

In [99]:
df_college2.sort_values("distance2")[
    [
        "AdmissionRate",
        "Undergraduates",
        "CarnegieClassification",
        "Ownership",
        "distance2"
    ]
].head(11)

,AdmissionRate,Undergraduates,CarnegieClassification,Ownership,distance2
Institution,,,,,
California Polytechnic State University-San Luis Obispo,0.3300,21090.0,Master's Colleges & Universities: Larger Programs,Public,0.000000
CUNY Hunter College,0.4590,17293.0,Master's Colleges & Universities: Larger Programs,Public,0.761441
CUNY Bernard M Baruch College,0.5056,15483.0,Master's Colleges & Universities: Larger Programs,Public,1.073601
CUNY John Jay College of Criminal Justice,0.4458,12834.0,Master's Colleges & Universities: Larger Programs,Public,1.184989
CUNY Brooklyn College,0.5136,12567.0,Master's Colleges & Universities: Larger Programs,Public,1.376321
California State Polytechnic University-Pomona,0.6062,26802.0,Master's Colleges & Universities: Larger Programs,Public,1.450297
CUNY Queens College,0.6078,14859.0,Master's Colleges & Universities: Larger Programs,Public,1.491386
CUNY Lehman College,0.5789,11609.0,Master's Colleges & Universities: Larger Programs,Public,1.660905
The University of West Florida,0.5281,8881.0,Master's Colleges & Universities: Larger Programs,Public,1.809410


**YOUR RESPONSE HERE**

I standardized the quantitative variables and used Euclidean distance, while also including Carnegie classification and ownership. The closest schools are public and have the same Carnegie classification as Cal Poly. Adding these categorical variables changed the matches and made them more similar in overall school type.

3\. The columns whose names begin with "PCIP" contain the proportions of students at each school studying various fields (e.g., Engineering, Psychology). Each field is represented by a two-digit code called the [CIP code](https://nces.ed.gov/ipeds/cipcode/browse.aspx?y=55).

If we only consider the proportions of students studying various fields, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [100]:
# YOUR CODE HERE. ADD CELLS AS 
pcip_vars = [col for col in df_college.columns if col.startswith("PCIP")]

In [101]:
pcip_data = df_college[pcip_vars]

In [102]:
cp_pcip = pcip_data.loc[school_name]

In [103]:
df_college["pcip_distance"] = np.sqrt(
    ((pcip_data - cp_pcip) ** 2).sum(axis=1)
)

In [104]:
df_college.sort_values("pcip_distance")[
    ["pcip_distance"]
].head(11)

,pcip_distance
Institution,
California Polytechnic State University-San Luis Obispo,0.000000
North Carolina State University at Raleigh,0.084343
Iowa State University,0.085054
University of Illinois Urbana-Champaign,0.111475
Mississippi State University,0.118799
Texas A & M University-College Station,0.123973
Clemson University,0.128333
Purdue University-Main Campus,0.134177
Virginia Polytechnic Institute and State University,0.137570


**YOUR RESPONSE HERE**

I used all PCIP variables and calculated Euclidean distance from Cal Poly. Since the variables are proportions on the same scale, I did not standardize them. The schools with the smallest distances have the most similar distributions of students across fields of study.